src/clients/c++/perf_client/load_manager.cc

```diff
@@ -27,21 +27,97 @@
+#include <cuda_runtime_api.h>
+
+#define RETURN_IF_CUDA_ERR(FUNC)                               \
+  {                                                            \
+    const cudaError_t result = FUNC;                           \
+    if (result != cudaSuccess) {                               \
+      return nic::Error(                                       \
+          ni::RequestStatusCode::INTERNAL,                     \
+          "CUDA exception (line " + std::to_string(__LINE__) + \
+              "): " + cudaGetErrorName(result) + " (" +        \
+              cudaGetErrorString(result) + ")");               \
+    }                                                          \
+  }
+
+#endif  // TRTIS_ENABLE_GPU
+
+namespace {
+
+#if TRTIS_ENABLE_GPU
+nic::Error
+CreateCUDAIPCHandle(
+    cudaIpcMemHandle_t* cuda_handle, void* input_d_ptr, int device_id = 0)
+{
+  // Set the GPU device to the desired GPU
+  RETURN_IF_CUDA_ERR(cudaSetDevice(device_id));
+
+  //  Create IPC handle for data on the gpu
+  RETURN_IF_CUDA_ERR(cudaIpcGetMemHandle(cuda_handle, input_d_ptr));
+
+  return nic::Error::Success;
+}
+#endif  // TRTIS_ENABLE_GPU
+
+void
+SerializeStringTensor(
+    std::vector<std::string> string_tensor, std::vector<char>* serialized_data)
+{
+  std::string serialized = "";
+  for (auto s : string_tensor) {
+    uint32_t len = s.size();
+    serialized.append(reinterpret_cast<const char*>(&len), sizeof(uint32_t));
+    serialized.append(s);
+  }
+
+  std::copy(
+      serialized.begin(), serialized.end(),
+      std::back_inserter(*serialized_data));
+}
+
+}  // namespace
+
 LoadManager::~LoadManager()
 {
-  early_exit = true;
-  // wake up all threads
-  wake_signal_.notify_all();
-
-  size_t cnt = 0;
-  for (auto& thread : threads_) {
-    thread.join();
-    if (!threads_stat_[cnt]->status_.IsOk()) {
-      std::cerr << "Thread [" << cnt
-                << "] had error: " << (threads_stat_[cnt]->status_)
+  nic::Error err;
+  if (shared_memory_type_ == SharedMemoryType::SYSTEM_SHARED_MEMORY) {
+    err = shared_memory_ctx_->UnregisterAllSharedMemory();
+    if (!err.IsOk()) {
+      std::cerr << "Unable to unregister all shared memory regions"
+                << std::endl;
+    }
+    for (auto region : shared_memory_regions_) {
+      err = UnmapSharedMemory(region.second, io_shm_size_[region.first]);
+      if (!err.IsOk()) {
+        std::cerr << "Unable to unmap shared memory with key (" << region.first
+                  << "): Starting: " << static_cast<void*>(region.second)
+                  << ", size: " << io_shm_size_[region.first] << std::endl;
+      }
+      err = UnlinkSharedMemoryRegion(region.first);
+      if (!err.IsOk()) {
+        std::cerr << "Unable to unlink shared memory with key: " << region.first
+                  << std::endl;
+      }
+    }
+  } else if (shared_memory_type_ == SharedMemoryType::CUDA_SHARED_MEMORY) {
+#if TRTIS_ENABLE_GPU
+    err = shared_memory_ctx_->UnregisterAllSharedMemory();
```

Comment:
This should be for GPU and CPU both. Shift the ENABLE GPU flag to after this

Reply:
Modified the structure.